python/lsst/pipe/tasks/matchBackgrounds.py

```diff
@@ -19,358 +19,473 @@
-            meanVar, meanVarErr = statObjIm.getResult(afwMath.VARIANCE)
-            meanBkgdLevel, meanBkgdLevelErr = statObjIm.getResult(afwMath.MEAN)
-            npoints, npointsErr = statObjIm.getResult(afwMath.NPOINT)
-            varList.append(meanVar)
-            meanBkgdLevelList.append(meanBkgdLevel)
-            coverageList.append(npoints)
-        if not coverageList:
-            raise pipeBase.TaskError(
-                "None of the candidate %s exist; cannot select best reference exposure" % (expDatasetType,))
-
-        # Normalize metrics to range from  0 to 1
-        varArr = numpy.array(varList)/numpy.nanmax(varList)
-        meanBkgdLevelArr = numpy.array(meanBkgdLevelList)/numpy.nanmax(meanBkgdLevelList)
-        coverageArr = numpy.nanmin(coverageList)/numpy.array(coverageList)
-
-        costFunctionArr = self.config.bestRefWeightVariance * varArr
-        costFunctionArr += self.config.bestRefWeightLevel * meanBkgdLevelArr
-        costFunctionArr += self.config.bestRefWeightCoverage * coverageArr
-        return numpy.nanargmin(costFunctionArr)
+        nx = warp.getWidth() // self.config.binSize
+        ny = warp.getHeight() // self.config.binSize
+
+        bgCtrl = BackgroundControl(nx, ny, self.statsCtrl, self.statsFlag)
+        bgCtrl.setUndersampleStyle(self.config.undersampleStyle)
+        warpBgMI = makeBackground(warp.getMaskedImage(), bgCtrl)
+
+        return warpBgMI, bgCtrl
+
+    def _fluxScale(self, exposure):
+        """Scales image to nJy flux using photometric calibration.
+
+        Parameters
+        ----------
+        exposure: `lsst.afw.image._exposure.ExposureF`
+            Exposure to scale.
+
+        Returns
+        -------
+        maskedImage: `lsst.afw.image._maskedImage.MaskedImageF`
+            Flux-scaled masked exposure.
+        """
+        maskedImage = exposure.getMaskedImage()
+        fluxZp = exposure.getPhotoCalib().instFluxToNanojansky(1)
+        exposure.image.array *= fluxZp
+
+        return maskedImage
 
     @timeMethod
     def matchBackgrounds(self, refExposure, sciExposure):
-        """Match science exposure's background level to that of reference exposure.
+        """Match science exposure's background level to that of reference
+        exposure.
 
-        Process creates a difference image of the reference exposure minus the science exposure, and then
-        generates an afw.math.Background object. It assumes (but does not require/check) that the mask plane
-        already has detections set. If detections have not been set/masked, sources will bias the
-        background estimation.
+        Process creates a difference image of the reference exposure minus the
+        science exposure, and then generates an afw.math.Background object. It
+        assumes (but does not require/check) that the mask plane already has
+        detections set. If detections have not been set/masked, sources will
```

Comment:
The whole point of background matching is that the astrophysical sources should subtract away, and I think that means we don't need to mask detections.  Even if the subtractions are't super clean, they should average out if we've done the right photometric scaling.